### PR TITLE
Redirect paths instead of discarding operations after move

### PR DESCRIPTION
**Pull request by jlelic**

### CODE REVIEW

**Review by martisekpetr — COMMENTED** [review]
I'd like to have test cases for the following scenarios:
* against remove - accepted op: remove row 0; proposed op: replace rows/0/column/0 -> should be removed
* against remove - accepted op: remove row 0; proposed op: replace rows/1/column/0 -> should shift indices
* against add - same two scenarios (add row, change module within/without row)

Basically, the tests for "add" and "remove" currently cover only operations on the same level, not nested.

File: src/index.ts [hunk 1]
```diff
@@ -151,6 +156,37 @@ function removeOperations<T extends Operation>(
...
+
+  for (const proposedOp of proposedOps) {
+    const matchesPathToFrom = proposedOp.path === acceptedFrom || proposedOp.path.indexOf(acceptedFrom + '/') === 0;
+
+    let matchesFromToFrom = false;
+    if (isFromOperation(proposedOp)) {
+      matchesFromToFrom = proposedOp.from === acceptedFrom || proposedOp.from.indexOf(acceptedFrom + '/') === 0;
+    }
+
+    const isProposedOpFrom = isFromOperation(proposedOp);
```
Comment (martisekpetr) [inline src/index.ts:167]: Move this 4 lines up and use in the `if`?

File: src/index.ts [hunk 2]
```diff
@@ -151,6 +156,37 @@ function removeOperations<T extends Operation>(
...
+
+    if (!isProposedOpFrom && matchesPathToFrom) {
+      proposedOp.path = acceptedPath + proposedOp.path.substr(acceptedFrom.length);
+      proposedOp.redirected = true;
+    } else if (isProposedOpFrom && matchesFromToFrom) {
+      const proposedFromOp = proposedOp as FromOperation;
+      proposedFromOp.from = acceptedPath + proposedFromOp.from.substr(acceptedFrom.length);
+      proposedFromOp.redirected = true;
+    }
+  }
```
Comment (martisekpetr) [inline src/index.ts:182]: What happens if `isProposedOpFrom && matchesPathToFrom`? E.g. proposed op is `{op: OpType.move, from: '/array/4', path: '/array/6'},` and accepted op is `{op: OpType.move, from: '/array/6', path: '/array/0'}`

Comment (jlelic) [reply src/index.ts:182]: It shifts the indices of the proposed op so the result is:
`{"from": "/array/5", "op": "move", "path": "/array/7"}`
I assume that's not correct, I'm not sure what the intended behavior should be though

Comment (martisekpetr) [reply src/index.ts:182]: Hm.. I guess that's fine, the `from` is transformed correctly. Ok, how about nested change: 
accepted: `{op: OpType.move, from: '/rows/6', path: '/rows/0'}`
proposed: `{op: OpType.move, from: '/rows/4/columns/0/modules/2', path: '/rows/6/columns/0/modules/1'}`? That should be transformed to `{op: OpType.move, from: '/rows/5/columns/0/modules/2', path: '/rows/0/columns/0/modules/1'}`

File: src/__tests__/new.test.ts [hunk 3]
```diff
@@ -189,20 +241,64 @@ describe('JSONPatchOT', () => {
...
 
-      // [0, 1, 2, 3, 4, 6]; <- Array after accepted copies
+      const proposedOps: Operation[] = [{op: OpType.replace, path: '/rows/0/columns/0/modules/2', value: 'Hello'}];
+
+      expect(JSONPatchOT(acceptedOps, proposedOps)).toEqual([
+        {op: OpType.replace, path: '/rows/5/columns/0/modules/2', value: 'Hello'},
+      ]);
+    });
+
+    it('redirect higher level paths after moving lower levels (moving to front)', () => {
```
Comment (jlelic) [inline src/__tests__/new.test.ts:280]: Added unit test(s) for nested changes

File: src/index.ts (hunk 2 above)
Comment (jlelic) [reply src/index.ts:182]: You're right, that wasn't working but it should be fixed now. I've also added this case and inverted one (moving from index 0 to 6) to the tests.

**Review by martisekpetr — APPROVED** [review]

**Pull request merged**